main.go

```diff
@@ -21,54 +21,193 @@
-	if len(*output) > 0 {
-		outFile, err = os.Create(*output)
-		defer outFile.Close()
+	if thriftRoot == "" {
+		thriftRoot, err = findCommonAncestor(module)
 		if err != nil {
 			log.Fatal(err)
 		}
+	} else {
+		thriftRoot, err = filepath.Abs(thriftRoot)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if err := verifyAncestry(module, thriftRoot); err != nil {
+			log.Fatal(err)
+		}
 	}
 
 	opts := gen.Options{
-		PackageName: filepath.Base(outDir), // TODO: customize using flags
-		Output:      outFile,
+		OutputDir:     outputDir,
+		PackagePrefix: packagePrefix,
+		ThriftRoot:    thriftRoot,
+		NoRecurse:     noRecurse,
 	}
+
 	if err := gen.Generate(module, &opts); err != nil {
 		log.Fatal(err)
 	}
 }
+
+// verifyAncestry verifies that the Thrift file for the given module and the
+// Thrift files for all imported modules are contained within the directory
+// tree rooted at the given path.
+func verifyAncestry(m *compile.Module, root string) error {
+	return m.Walk(func(m *compile.Module) error {
+		path, err := filepath.Rel(root, m.ThriftPath)
+		if err != nil {
+			return fmt.Errorf(
+				"could not resolve path for %q: %v", m.ThriftPath, err)
+		}
+
+		if strings.HasPrefix(path, "..") {
+			return fmt.Errorf(
+				"%q is not contained in the %q directory tree",
+				m.ThriftPath, root)
+		}
+
+		return nil
+	})
+}
+
+// findCommonAncestor finds the deepest common ancestor for the Thrift file
+// for the given module and all the modules imported by it.
+func findCommonAncestor(m *compile.Module) (string, error) {
+	var result []string
+	var lastString string
+
+	err := m.Walk(func(m *compile.Module) error {
+		thriftPath := m.ThriftPath
+		if !filepath.IsAbs(thriftPath) {
+			return fmt.Errorf(
+				"ThriftPath must be absolute: %q is not absolute", thriftPath)
+		}
+		thriftPath = filepath.Dir(thriftPath)
+
+		// SplitAfter("/foo/bar", "/") = ["/", "foo/", "bar"]
+		parts := strings.SplitAfter(thriftPath, string(filepath.Separator))
+		if result == nil {
+			result = parts
+			lastString = thriftPath
+			return nil
+		}
+
+		result = commonPrefix(result, parts)
+		if len(result) == 0 {
+			return fmt.Errorf(
+				"%q does not have a prefix in common with %q",
+				thriftPath, lastString)
+		}
+
+		lastString = thriftPath
+		return nil
+	})
+
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(result...), nil
+}
+
+// commonPrefix finds the shortest common prefix for the two lists.
+//
+// An empty slice may be returned if the two lists don't have a common prefix.
+func commonPrefix(l, r []string) []string {
+	var i int
+	for i = 0; i < len(l) && i < len(r); i++ {
+		if l[i] != r[i] {
+			break
+		}
+	}
+	return l[:i]
+}
+
+// determinePackagePrefix determines the package prefix for Go packages
+// generated in this file.
+//
+// dir must be an absolute path.
+func determinePackagePrefix(dir string) (string, error) {
+	gopath := os.Getenv("GOPATH")
+	if gopath == "" {
+		return "", errors.New("$GOPATH is not set")
+	}
+
```

Comment:
GOPATH could have multiple colon separated paths, so split on `:` and iterate over all of them and check for at least one match.